pkg/bloombuild/planner/planner_test.go

```diff
@@ -377,77 +384,237 @@ func Test_BuilderLoop(t *testing.T) {
 		},
 	}
 
-	// Create planner
-	planner, err := New(cfg, limits, schemaCfg, storageCfg, storage.NewClientMetrics(), nil, logger, prometheus.DefaultRegisterer)
+	reg := prometheus.NewPedanticRegistry()
+	planner, err := New(cfg, limits, schemaCfg, storageCfg, storage.ClientMetrics{}, nil, logger, reg)
 	require.NoError(t, err)
 
-	// Start planner
-	err = services.StartAndAwaitRunning(context.Background(), planner)
-	require.NoError(t, err)
-	t.Cleanup(func() {
-		err := services.StopAndAwaitTerminated(context.Background(), planner)
-		require.NoError(t, err)
-	})
+	return planner
+}
 
-	// Enqueue tasks
-	for i := 0; i < nTasks; i++ {
-		task := NewTask(
-			context.Background(), time.Now(),
-			protos.NewTask(config.NewDayTable(config.NewDayTime(0), "fake"), "fakeTenant", v1.NewBounds(0, 10), tsdbID(1), nil),
-		)
+func Test_BuilderLoop(t *testing.T) {
+	const (
+		nTasks    = 100
+		nBuilders = 10
+	)
 
-		err = planner.enqueueTask(task)
-		require.NoError(t, err)
-	}
+	for _, tc := range []struct {
+		name                     string
+		limits                   Limits
+		expectedBuilderLoopError error
+
+		// modifyBuilders should leave the builders in a state where they will not return or return an error
+		modifyBuilders func(builders []*fakeBuilder)
+		// resetBuilders should reset the builders to a state where they will return no errors
+		resetBuilders func(builders []*fakeBuilder)
+	}{
+		{
+			name:                     "success",
+			limits:                   &fakeLimits{},
+			expectedBuilderLoopError: errPlannerIsNotRunning,
+		},
+		{
+			name:                     "error rpc",
+			limits:                   &fakeLimits{},
+			expectedBuilderLoopError: errPlannerIsNotRunning,
+			modifyBuilders: func(builders []*fakeBuilder) {
+				for _, builder := range builders {
+					builder.SetReturnError(true)
+				}
+			},
```

Comment:
nit: since you always modify all builds in a loop, you could simply provide the function modify a single builder

```suggestion
			modifyBuilder: func(builder *fakeBuilder) {
				builder.SetReturnError(true)
			},
```

Reply:
Done. Thank you!